### PR TITLE
ArgumentParser: gdal_calc + others

### PR DESCRIPTION
**Pull request by idanmiara**
## What does this PR do?

gdal_argparse.py - GDALArgumentParser an extended `argparse.ArgumentParser` to be used with all `gdal-utils`
gdal_calc.py - use `GDALArgumentParser ` instead of the old `optparse.OptionParser` 
use minor improvements to `ArgumentParser` uses in other utils

### CODE REVIEW

File: gdal/swig/python/gdal-utils/osgeo_utils/auxiliary/gdal_argparse.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            # extend was introduced to the stdlib in Python 3.8
+            self.register('action', 'extend', ExtendAction)
+
+    def parse_args(self, args=None, optfile_arg=None, **kwargs):
+        if (args is not None
+           and optfile_arg in args
+           and self.fromfile_prefix_chars is not None
+           and optfile_arg is not None):
+
+            # Replace '--optfile x' with the standard '@x'
```
Comment (rouault) [inline gdal/swig/python/gdal-utils/osgeo_utils/auxiliary/gdal_argparse.py:58]: I'm not sure they are completely equivalent. From what I can read in https://docs.python.org/3/library/argparse.html#fromfile-prefix-chars , @x requires one argument per line whereas --optfile x can support potentially arguments on the same line

Comment (idanmiara) [reply gdal/swig/python/gdal-utils/osgeo_utils/auxiliary/gdal_argparse.py:58]: Thanks for your comment.
I actually handled that in GDALArgumentaParser. 
It passes the relevant test7 for that cases you've mentioned.

Comment (rouault) [reply gdal/swig/python/gdal-utils/osgeo_utils/auxiliary/gdal_argparse.py:58]: ok, any reason to have optfile_arg default to None and being set only in gdal_calc.py to "--optfile" ?

And is this code exercised by the test suite ? (my mantra is "non-tested code is broken, or will soon be broken", and it is even more true for an interpreted language like Python)

Comment (idanmiara) [reply gdal/swig/python/gdal-utils/osgeo_utils/auxiliary/gdal_argparse.py:58]: --optfile is for backwards compatibility. I can't see a reason to encourage people to use it ovet the standard pefix@.
I'll set it on any until that currently have this option.

I've learned to know your mantra :) 
Almost all the tests of `gdal_calc` use the ArgParser so it is tested there through it.
Did you refer to any other code?

Comment (rouault) [reply gdal/swig/python/gdal-utils/osgeo_utils/auxiliary/gdal_argparse.py:58]: Regarding testing, I refer to this new code replacing --optfile by ``@ ``

Comment (idanmiara) [reply gdal/swig/python/gdal-utils/osgeo_utils/auxiliary/gdal_argparse.py:58]: Calc test 7 tests that as it inputs `- -optfile` and the parser transforms it to `@` and parses it as the normal way. 

I can add a test for input with the prefix.

Comment (idanmiara) [reply gdal/swig/python/gdal-utils/osgeo_utils/auxiliary/gdal_argparse.py:66]: I've added tests for `@` prefix and a deprecation warning for `--optfile`.

**Pull request merged**